maps-compose/src/main/res/values/ids.xml

```diff
@@ -1,4 +1,9 @@
 <?xml version="1.0" encoding="utf-8"?>
 <resources>
     <item name="maps_compose_nodraw_container_view" type="id" />
+
+    <item name="maps_compose_map_view_tag_composition" type="id" />
+    <item name="maps_compose_map_view_tag_click_listeners" type="id" />
+    <!-- For test purposes -->
+    <item name="maps_compose_map_view_tag_debug_id" type="id" />
```

Comment:
I would use hard-coded constants here. (E.g. pick some random numbers.) Having the build pick numbers would make any potential id conflicts with other components intermittent instead of predictable.

Reply:
If you want to store things in the MapView's tag using keys you have to use an ID resource, you cannot use a normal constant. 
There is also just a `tag` without an ID, and then you don't need a key, but that makes it slightly more complex because you'd need a wrapper object, which probably would require synchronization, but that's totally doable.

Reply:
I'll get back to you regarding the other issues later :)

Reply:
> If you want to store things in the MapView's tag using keys you have to use an ID resource, you cannot use a normal constant.

The Android framework does not know whether you're using an ID resource or just some Int constant; resource IDs are internal to your app. The actual Int IDs are generated during app build, not during the library build, so they differ from app to app and potentially build to build. This would make it hard to reproduce an issue with colliding IDs if MapView or AndroidView were to store data of their own in the tag. Easier to reproduce problems with normal Int constants in this case.

> There is also just a `tag` without an ID, and then you don't need a key, but that makes it slightly more complex because you'd need a wrapper object, which probably would require synchronization, but that's totally doable.

Using a data type in this manner may be preferable if all data is generated, updated, and stored at the same time; you can use immutable objects in this case. If stored at different times you might prefer to store with individual IDs.

Reply:
Right.
I tested to use my own constants for keys and it didn't throw an exception for some reason but it's best to not do that because it might still cause crashes on other devices or in other scenarios.
For now I created a MapTagData data class with mutable fields, stored in the default tag. Would you prefer immutable fields instead?
I'm not sure what's the best approach.
If using immutable fields it would be good to synchronize it for the future if additional fields are added.

Reply:
Right now it looks like you would not need the tag stuff anymore once you get rid of special ComponentCallback handling and debug things.

Otherwise I'd typically stick with an immutable data class & copy constructor unless it's undesirable for a specific reason. Simple mutable types are just a bit of a red flag these days.

Reply:
Don't add explicit synchronization unless it's actually needed, which should be rare. There are many ways to keep this code safe.

Reply:
I should add: anything that can be made private to just the AndroidView should be stored in the tag. It's local, so store it locally. Kinda critical to insulate local data from composition, as composition & remember lifecycle is different from reusable AndroidView lifecycle.

I have not looked around the code for this specifically, but I suspect that there is data to which this applies.

Thinking about it, what I said earlier about mutable vs. immutable may not be right. If our data structure in the tag is merely a bucket for storing different unrelated kinds of data then it does not generally make sense to make the bucket immutable, as long as the bucket is initialized & set exactly once, and disposed exactly once. Each piece of data inside the bucket should be immutable.

Reply:
I still think the `ComponentCallbacks` needs to be stored in the MapView's tag because it has to be registered during the MapView's entire lifecycle and then finally unregistered. Meanwhile it has to be stored somewhere.